Python package fixes

pyrox0 opened the pull request:
Fixes a few packages that are broken on python 3.12, mainly related to their usage of `nose`, by disabling tests or otherwise migrating them to `pytestCheckHook`.
This also migrates binwalk to the supported fork at OSPG/binwalk, which also fixes a `nose` issue by it migrating to `pytest`.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

layus reviewed (approved):
Looks okay to me. I do not understand why you have taken on you to change the `with lib;` from meta, but I am not one to bikeshed on these things.

Thanks for the work.
I will wait a bit to let others review it too. Do not hesitate to ping me if it is not merged withing the next few days.

natsukium commented on pkgs/development/python-modules/citeproc-py/default.nix:
Please replace `format = setuptools` with `pyproject = true`.
Also, replace `sha256` with `hash`.

pyrox0 replied:
Done

natsukium reviewed (changes requested):
I recommend submitting one PR per package next time.

In #311054, hexa says
> On top of that, packages relying on nose have largely been unmaintained and might be dead leafs or using another test framework entirely. It is also to be expected that they need to be converted to the PEP517 builder in the process of fixing them.

This indicates that we need to change the packages significantly.

pyrox0 commented on pkgs/development/python-modules/capstone/default.nix:
```diff
@@ -11,8 +11,8 @@ buildPythonPackage rec {
-  # distutils usage
-  disabled = pythonAtLeast "3.12";
+  # Remove distutils usage when > 3.12 as 3.12 removes distutils.
+  patches = if (pythonAtLeast "3.12") then [ ./remove-distutils.patch ] else [];
```
Removed the condition, and submitted upstream as https://github.com/capstone-engine/capstone/pull/2369

pyrox0 commented on pkgs/development/python-modules/cgroup-utils/default.nix:
```diff
@@ -1,30 +1,34 @@
-    sha256 = "0qnbn8cnq8m14s8s1hcv25xjd55dyb6yy54l5vc7sby5xzzp11fq";
+    hash = "sha256-2IVw/+/FL33YLpQU783yrZQmexGbwaCRJqEibBmyy2I=";
   };
 
```
Added the importCheck, but tests don't work as they need to access the filesystem(`/sys`) when the module is imported, and therefore don't work in Nix's environment. I set `doCheck = false;` and added a comment explaining why.

jonringer commented on pkgs/development/python-modules/citeproc-py/default.nix:
```diff
@@ -2,39 +2,41 @@
+  # doCheck = false;  # seems to want a Git repository, but fetchgit with leaveDotGit also fails
   pythonImportsCheck = [ "citeproc" ];
 
-  meta = with lib; {
```
just curious, is this the new best practice?

pyrox0 replied:
Yes, see #208242.

jonringer replied:
I'm kind of aligned with https://github.com/NixOS/nixpkgs/issues/208242#issuecomment-1700396671. File-wide with usage is an anti pattern, but scoped to something like a meta block is probably harmless. Either way, don't have a strong preference as long we don't flip flop back and forth.

pyrox0 replied:
I've done it in a bunch of previous PRs and no one's been opposed to it, plus there's no reason to change it back since while not needed here, this is technically better. :shrug:

natsukium commented on pkgs/development/python-modules/blessings/default.nix:
```diff
@@ -3,33 +3,35 @@
   fetchPypi,
   six,
-  nose,
+  pynose,
```
`pynose` violates the license and will be removed in https://github.com/NixOS/nixpkgs/pull/325669
If you need `nose` for testing, please disable tests in 3.12.

pyrox0 replied:
Will remove pynose from all applicable packages.

pyrox0 committed:
python312Packages.citeproc-py: nose -> pytest

pyrox0 committed:
python312Packages.chart-studio: nose -> pytest

pyrox0 committed:
python312Packages.cgroup-utils: nose -> tests disabled

They require privilaged filesystem access(/proc), so they do not run in
Nix's sandbox.

pyrox0 committed:
python312Packages.blessings: nose -> disabled tests

See upstream PR #165 for details

pyrox0 committed:
python312Packages.binwalk: 2.3.4 -> 2.4.1

Updates to the community supported fork of binwalk, and modernizes the
code to modern Nix Python packaging standards.

pyrox0 commented:
This previously had changes related to Capstone's Python bindings. Those have been dropped as out-of-scope for this PR, and can be resubmitted later once upstream fully drops distutils in a later release.

pyrox0 commented:
merge conflict fixed.

7c6f434c commented on pkgs/development/python-modules/cgroup-utils/default.nix:
```diff
@@ -1,30 +1,34 @@
     sed -i -e "/argparse/d" setup.py
+    substituteInPlace setup.py --replace-fail \
+    "tests_require=['nose', 'pep8']" "tests_require=['pynose', 'pep8']"
+    substituteInPlace test_pep8.py --replace-fail \
+    'print message' 'print(message)'
```
Do you fix then later remove the same file in the current version?

pyrox0 commented:
Closing since this has been superseded by multiple other PRs and it's no longer needed.
